src/slow_bruteforce_interpreter.rs

```diff
@@ -0,0 +1,1016 @@
+    ///
+    /// NB: Requires a second validation pass to produce a proper SPPF.
+    #[derive(Default, Debug)]
+    struct ApproxForest {
+        possibilities: HashMap<(IRule, RangeFrom<usize>), ApproxData<usize>>,
+    }
+
+    #[derive(Debug)]
+    enum ApproxData<T> {
+        One(T, Option<RangeTo<usize>>),
+        Many(BTreeSet<T>),
+    }
+
+    impl<T> ApproxData<T> {
+        fn iter<'a>(&'a self) -> impl Iterator<Item = &'a T> {
+            match self {
+                ApproxData::One(x, _) => Some(x).into_iter().chain(None.into_iter().flatten()),
+                ApproxData::Many(set) => None
+                    .into_iter()
+                    .chain(Some(set.iter()).into_iter().flatten()),
+            }
+        }
+    }
+
+    impl ApproxForest {
+        fn add(&mut self, key: (IRule, RangeFrom<usize>), x: usize, end: Option<RangeTo<usize>>) {
+            use std::collections::hash_map::Entry;
+
+            match self.possibilities.entry(key) {
+                Entry::Vacant(entry) => {
+                    entry.insert(ApproxData::One(x, end));
+                }
+                Entry::Occupied(entry) => {
+                    let data = entry.into_mut();
+                    match data {
+                        ApproxData::One(y, old_end) if x == *y => {
+                            if end != *old_end {
+                                *old_end = None;
+                            }
+                        }
+                        ApproxData::One(y, _) => {
+                            *data = ApproxData::Many([x, *y].iter().copied().collect());
+                        }
+                        ApproxData::Many(set) => {
+                            set.insert(x);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn parse_inner<'i, Pat: Clone + Ord + Hash + fmt::Debug, I: Input>(
+        // FIXME(eddyb) group some of these in a `struct`.
+        cx: &Context<Pat>,
+        grammar: &crate::Grammar,
+        parser: &RefCell<Parser<'_, 'i, SlowBruteforceInterpreter<'_, Pat>, I, Pat>>,
+        parse_cached: &mut dyn FnMut((IRule, Range<'i>)) -> Rc<CachedParse>,
+        approx_forest: &mut ApproxForest,
+        rule: IRule,
+        range: Range<'i>,
+    ) -> BTreeSet<usize>
+    where
+        I::Slice: InputMatch<Pat>,
+    {
+        match cx[rule] {
+            Rule::Empty => iter::once(0).collect(),
+            // FIXME(eddyb) find a way to avoid cloning the pattern.
+            Rule::Eat(ref pat) => parser
+                .borrow_mut()
+                .with_result_and_remaining(Range(range.frontiers().0), range)
+                .input_consume_left(pat.clone())
+                .map(|parser| parser.result().len())
+                .into_iter()
+                .collect(),
+            // FIXME(eddyb) avoid cloning the set from behind a `Rc`.
+            // May require something like `Cow` but for `T | Rc<T>`?
```

Comment:
Wouldn't this be [`Rc::try_unwrap`](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.try_unwrap)?

Reply:
No, I want to iterate the set without ever cloning it.